packages/studio-base/src/players/UserNodePlayer/nodeTransformerWorker/generateDataSourceLib.ts

```diff
@@ -0,0 +1,172 @@
+  ["float64", "number"],
+  ["int64", "number"],
+  ["uint64", "number"],
+  ["string", "string"],
+  ["bool", "boolean"],
+  ["time", "Time"],
+  ["duration", "Duration"],
+]);
+
+export const typedArrayMap = new Map<string, string>([
+  ["uint8", "Uint8Array"],
+  ["int8", "Int8Array"],
+  ["int16", "Int16Array"],
+  ["uint16", "Uint16Array"],
+  ["int32", "Int32Array"],
+  ["uint32", "Uint32Array"],
+  ["int64", "BigInt64Array"],
+  ["uint64", "BigUint64Array"],
+  ["float32", "Float32Array"],
+  ["float64", "Float64Array"],
+]);
+
+function safeString(str: string): string {
+  return JSON.stringify(str) as string;
+}
+
+export const generateTypesInterface = (datatypes: RosDatatypes): string => {
+  const seenDatatypes = new Set();
+  let src = `
+    /**
+     * DataSourceTypes enumerates the message type definitions for for all the types
+     * in the current data source.
+     *
+     * Use subtype notation to access the type by schema name:
+     *
+     * \`\`\`
+     * type Msg = DataSourceTypes["geometry_msgs/Pose"];
+     * type Msg = DataSourceTypes["pkg.geometry.Pose"];
+     * \`\`\`
+     */
+    export type DataSourceTypes = {
+  `;
+
+  for (const [datatype, definition] of datatypes) {
+    // Avoid adding a repeating datatype name again.
+    // We shouldn't have this happen so we add a comment indicating it happened.
+    if (seenDatatypes.has(datatype)) {
+      src += `\n// ${datatype} appeared multiple times in datatypes`;
+      continue;
+    }
+    seenDatatypes.add(datatype);
+
+    src += `\n${safeString(datatype)}: {`;
+
+    for (const field of definition.definitions) {
+      const { type, isConstant, isArray } = field;
+      const typedArray = typedArrayMap.get(type);
+      const rosPrimitive = rosPrimitiveToTypescriptTypeMap.get(type);
+
+      const fieldName = safeString(field.name);
+
+      if (isConstant === true) {
+        src += `\n // ${field.name} = ${field.valueText}`;
+      } else if (isArray === true) {
+        if (typedArray) {
+          src += `\n${fieldName}: ${typedArray},`;
+        } else if (rosPrimitive) {
+          src += `\n${fieldName}: ${rosPrimitive}[],`;
+        } else {
+          src += `\n${fieldName}: DataSourceTypes[${safeString(type)}][],`;
+        }
+      } else {
+        if (rosPrimitive) {
+          src += `\n${fieldName}: ${rosPrimitive},`;
+        } else {
+          src += `\n${fieldName}: DataSourceTypes[${safeString(type)}],`;
+        }
+      }
+    }
+
+    src += "\n},";
+  }
+
+  src += "\n};";
+
+  return src;
+};
+
+function generateTypesByTopicInterface(topics: Topic[]): string {
+  let src = `
+    /**
+     * MessageEventByTopic enumerates the MessageEvent types for all the topics in
+     * the current data source.
+     *
+     * Use subtype notation to access the MessageEvent type by topic name:
+     *
+     * \`\`\`
+     * type MsgEventFoo = MessageEventByTopic["/foo"];
+     * \`\`\`
+     */
+    export type MessageEventByTopic = {`;
```

Comment:
We should try to be consistent with pluralization between DataSourceTypes & here.

```suggestion
    export type MessageEventsByTopic = {`;
```

Reply:
I would probably change `DataSourceTypes` to `DataSourceType` then. The type from `MessageEventByTopic` is a single MessageEvent not an array. Maybe thats not a distinction that matters?